test/gtest/error_handling.cpp

```diff
@@ -0,0 +1,338 @@
+                                     nixl_xfer_dlist_t& sReq_descs,
+                                     nixl_xfer_dlist_t& rReq_descs,
+                                     nixlXferReqH*& req_handle) const {
+    nixl_opt_args_t extra_params = { .backends = {m_backend} };
+    extra_params.notifMsg        = "notification";
+    extra_params.hasNotif        = true;
+    return m_priv->createXferReq(op, sReq_descs, rReq_descs, m_MetaRemote,
+                                 req_handle, &extra_params);
+}
+
+nixl_status_t
+TestErrorHandling::Agent::postXferReq(nixlXferReqH* req_handle) const {
+    return m_priv->postXferReq(req_handle);
+}
+
+nixl_status_t
+TestErrorHandling::Agent::waitForCompletion(nixlXferReqH* req_handle) {
+    nixl_status_t status;
+
+    do {
+        status = m_priv->getXferStatus(req_handle);
+    } while (status == NIXL_IN_PROG);
+
+    m_priv->releaseXferReq(req_handle);
+    return status;
+}
+
+nixl_status_t
+TestErrorHandling::Agent::waitForNotif(const std::string& expectedNotif) {
+    nixl_notifs_t notif_map;
+
+    do {
+        EXPECT_EQ(NIXL_SUCCESS, m_priv->getNotifs(notif_map));
+    } while (notif_map.empty());
+
+    std::vector<std::string> notifs = notif_map[m_MetaRemote];
+    EXPECT_EQ(1,             notifs.size());
+    EXPECT_EQ(expectedNotif, notifs.front());
+    return NIXL_SUCCESS;
+}
+
+void TestErrorHandling::Agent::fillData() {
+    m_mem.fillData();
+}
+
+bool TestErrorHandling::Agent::dataCmp(const TestErrorHandling::Agent& other) const {
+    return m_mem.m_data == other.m_mem.m_data;
+}
+
+TestErrorHandling::TestErrorHandling() {
+    const char* plugin_dir_env = getenv("NIXL_PLUGIN_DIR");
+    m_plugin_dir_backup.first  = plugin_dir_env != nullptr;
+    if (m_plugin_dir_backup.first) {
+        m_plugin_dir_backup.second = plugin_dir_env;
+    }
+
+    // Set up test environment
+    // Load plugins from build directory
+    std::string plugin_dir = std::string(BUILD_DIR) + "/src/plugins/ucx";
+    setenv("NIXL_PLUGIN_DIR", plugin_dir.c_str(), 1);
+
+    std::cout << "set NIXL_PLUGIN_DIR: " << getenv("NIXL_PLUGIN_DIR")
+              << std::endl;
+}
+
+TestErrorHandling::~TestErrorHandling() {
+    if (m_plugin_dir_backup.first) {
+        setenv("NIXL_PLUGIN_DIR", m_plugin_dir_backup.second.c_str(), 1);
+        std::cout << "restore NIXL_PLUGIN_DIR: " << getenv("NIXL_PLUGIN_DIR")
+                  << std::endl;
+    } else {
+        std::cout << "unset NIXL_PLUGIN_DIR" << std::endl;
+        unsetenv("NIXL_PLUGIN_DIR");
+    }
+}
+
+template<TestErrorHandling::TestType test_type, enum nixl_xfer_op_t op>
+void TestErrorHandling::testXfer() {
+    m_Initiator.init("initiator");
+    m_Target.init("target");
+
+    exchangeMetaData();
+
+    for (size_t i = 0; i < numIter<test_type>(); ++i) {
+        nixlXferReqH* req_handle = postXfer(op, isFailure<test_type>(i));
+        nixl_status_t status     = m_Initiator.waitForCompletion(req_handle);
+
```

Comment:
minor: could use auto type

Reply:
I don't like `auto` for simple types, in many cases it disimproves readability